### ISSUE TITLE
Access-Control-Allow-Origin

### ISSUE THREAD

**Issue by zhanzhenzhen** [fixerAPI/fixer#18]
Could you consider adding a `Access-Control-Allow-Origin` header to the response so we can call it on browser?

**Comment by amitmtrn** [comment]
+1

**Comment by hakanensari** [comment]
@zhanzhenzhen @amitmtrn,

Sorry for not responding earlier. We should already be handling cross-origin. See output below:

``` sh
$ curl -I -H "ORIGIN: http://127.0.0.1" http://api.fixer.io
HTTP/1.1 200 OK
Server: nginx/1.9.14
Date: Fri, 08 Apr 2016 18:02:23 GMT
Content-Type: application/json
Content-Length: 49
Connection: keep-alive
Access-Control-Allow-Origin: http://127.0.0.1
Access-Control-Allow-Methods: POST, GET, OPTIONS
Access-Control-Allow-Headers: *, Content-Type, Accept, AUTHORIZATION, Cache-Control
Access-Control-Allow-Credentials: true
Access-Control-Max-Age: 1728000
Access-Control-Expose-Headers: Cache-Control, Content-Language, Content-Type, Expires, Last-Modified, Pragma
X-Content-Type-Options: nosniff
```

This is also covered in the specs [here](https://github.com/hakanensari/fixer-io/blob/master/spec/api_spec.rb#L53-L56).

**Comment by jsmmth** [comment]
@hakanensari 

`XMLHttpRequest cannot load http://api.fixer.io/latest. The 'Access-Control-Allow-Origin' header has a value 'http://fixer.io' that is not equal to the supplied origin. Origin '{my_website_url}' is therefore not allowed access.`

**Comment by zhanzhenzhen** [comment]
In Chrome, visit:

http://api.fixer.io/latest?base=USD

I can't see the `Access-Control-Allow-Origin` header.

**Comment by jsmmth** [comment]
Yeah I can view it in the browser fine, and infact I can from postman which is strange. But when I try to access it from a json request it seems to fail due to access control. Anyone worked out a solution?

**Comment by jsmmth** [comment]
For anyone struggling with this make sure you request with https:// not http:// that fixed the issue for me

**Comment by hakanensari** [comment]
We should now be returning the `Access-Control-Allow-Origin` header to all requests. For [good or worse](https://github.com/britg/sinatra-cross_origin/blob/master/lib/sinatra/cross_origin.rb#L29), the app was previously expecting an `HTTP_ORIGIN` header to do.

Let me know if this solves your issues.

**Comment by zhanzhenzhen** [comment]
Yes, solved! Thank you!

**Comment by visageae** [comment]
Does not work with pre-flight OPTIONS requests 

```
curl -I -X OPTIONS -H "ORIGIN: http://127.0.0.1" http://api.fixer.io
HTTP/1.1 404 Not Found
Server: nginx/1.9.15
Date: Mon, 25 Apr 2016 10:10:23 GMT
Content-Type: text/html;charset=utf-8
Content-Length: 21
Connection: keep-alive
Allow: HEAD, GET, OPTIONS
Access-Control-Allow-Headers: X-Requested-With, X-HTTP-Method-Override, Content-Type, Cache-Control, Accept
X-Cascade: pass
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN

```

**Comment by hakanensari** [comment]
@visageae: should work now.

**Comment by visageae** [comment]
Not yet,
Now Access-Control-Allow-Origin headers are not present in responses for OPTIONS requests.
Can you add that ?

Thanks

**Comment by jsmmth** [comment]
For anyone with this issue I changed from http to https and it worked :) make sure you're doing the same.

**Comment by zhanzhenzhen** [comment]
Seems it returns back to the old behavior?

@hakanensari: Is it possible to include a `origin` header in browser request? If it's possible then it has no problem.

I know some services don't require the `origin` header, for example:
https://query.yahooapis.com/v1/public/yql?q=select%20*%20from%20yahoo.finance.quotes%20where%20symbol%20in%20(%22AAPL%22)&env=store%3A%2F%2Fdatatables.org%2Falltableswithkeys&format=json

**Comment by zhanzhenzhen** [comment]
Sorry, just found `origin` header is automatically added while using xhr in browser.

**Comment by hakanensari** [comment]
@zhanzhenzhen, 

It seems the request has to include an `Origin` header. Otherwise, we're not supposed to treat it as an access control request.

To quote from the [MDN website](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#The_HTTP_request_headers):

> Note that in any access control request, the ORIGIN header is always sent.

I'm assuming this explains the behaviour of the [library](https://github.com/cyu/rack-cors) that I included in the project today.

Long story short, does adding this header to your request resolve the issue on your side?

**Comment by zhanzhenzhen** [comment]
Now I have no problem, thanks. I found `origin` is automatically added while using XMLHttpRequest (not directly typing in the address bar) to send the request. Browser prohibits manually setting this header.

**Comment by visageae** [comment]
> Now Access-Control-Allow-Origin headers are not present in responses for OPTIONS requests.

Still no Access-Control headers in response headers thus preventing some frameworks to send further requests. Here is the only response I get sending an OPTIONS request :

```
Request URL:https://api.fixer.io/latest
Request Method:OPTIONS
Status Code:200 
Remote Address:178.62.197.164:443
```

```
content-length:0
content-type:text/html;charset=utf-8
date:Sun, 05 Jun 2016 12:35:14 GMT
server:nginx/1.9.15
status:200
vary:Origin
x-content-type-options:nosniff
x-frame-options:SAMEORIGIN
x-xss-protection:1; mode=block
```

Responses should contain sth like : 

```
Access-Control-Allow-Headers:accept, authorization
Access-Control-Allow-Methods:GET,HEAD,PUT,PATCH,POST,DELETE
Access-Control-Allow-Origin:*
```

Do you plan to fix that ?
Thank you very much

**Comment by hakanensari** [comment]
@visageae,

It does return the headers:

``` bash
curl -I -X OPTIONS -H "ORIGIN: *" http://api.fixer.io
HTTP/1.1 200 OK
Server: nginx/1.9.15
Date: Mon, 06 Jun 2016 06:49:52 GMT
Content-Type: text/html;charset=utf-8
Content-Length: 0
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET
Access-Control-Expose-Headers: 
Access-Control-Max-Age: 1728000
Access-Control-Allow-Credentials: true
Vary: Origin
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
```

**Comment by visageae** [comment]
Hum very strange. 
This curl is fine : 

```
curl 'https://api.fixer.io/latest' -X OPTIONS -H "ORIGIN: *" -H 'access-control-request-method: GET' -v
HTTP/1.1 200 OK
Server: nginx/1.9.15
Date: Mon, 06 Jun 2016 12:08:20 GMT
Content-Type: text/plain
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET
Access-Control-Expose-Headers: 
Access-Control-Max-Age: 1728000
Access-Control-Allow-Credentials: true
X-Content-Type-Options: nosniff
```

But if you add the access-control-request-headers header, it stops working.
Example : 

```
curl 'https://api.fixer.io/latest' -X OPTIONS -H "ORIGIN: *" -H 'access-control-request-method: GET' -H 'access-control-request-headers: accept' -v
HTTP/1.1 200 OK
Server: nginx/1.9.15
Date: Mon, 06 Jun 2016 12:10:05 GMT
Content-Type: text/html;charset=utf-8
Content-Length: 0
Connection: keep-alive
Vary: Origin
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
```

You should consider adding this header to your responses to preflight requests : 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Access-Control-Allow-Headers

**Comment by hakanensari** [comment]
I'm afraid your curl query is bad.You can't provide multiple header parameters: last one overrides the previous.

**Comment by visageae** [comment]
Uh? So my whole life is a lie? This is not how curl works. 

From curl man page. 

-H, --header <header>
(HTTP) Extra header to include in the request when sending HTTP to a server. You may specify any number of extra headers.

**Comment by hakanensari** [comment]
@visageae, disregard my earlier comment. You're right. The API should now be returning the correct headers.

**Comment by visageae** [comment]
Amazing !
Now this is officially angular ready ! :)
Thank you
